File: android/src/main/java/com/nitroplay/video/core/ListenerRegistry.kt
package com.nitroplay.video.core

import com.margelo.nitro.video.ListenerSubscription
import java.util.UUID

data class ListenerEntry(val id: UUID, val eventName: String, val callback: Any)

class ListenerRegistry {
  private val entries = mutableListOf<ListenerEntry>()
  private val lock = Any()

  fun <T> add(event: String, listener: T): ListenerSubscription {
    val id = UUID.randomUUID()
    synchronized(lock) { entries.add(ListenerEntry(id, event, listener as Any)) }
    return ListenerSubscription {
      synchronized(lock) { entries.removeAll { it.id == id } }
    }
  }

  @Suppress("UNCHECKED_CAST")
  fun <T> emit(event: String, invoke: (T) -> Unit) {
    val snapshot = synchronized(lock) { entries.filter { it.eventName == event }.toList() }
    for (entry in snapshot) {
      try {
        invoke(entry.callback as T)
      } catch (e: Exception) {
        android.util.Log.e("ListenerRegistry", "Error in $event listener", e)
      }
    }
  }

  fun clearAll() {
    synchronized(lock) { entries.clear() }
  }

  fun hasListeners(event: String): Boolean {
    return synchronized(lock) { entries.any { it.eventName == event } }
  }
}

File: ios/core/ListenerRegistry.swift
import Foundation
import NitroModules

struct ListenerEntry {
  let id: UUID
  let eventName: String
  let callback: Any
}

final class ListenerRegistry {
  private var listeners: [ListenerEntry] = []
  private let lock = NSLock()

  func add<T>(event: String, listener: T) -> ListenerSubscription {
    let id = UUID()
    lock.lock()
    listeners.append(ListenerEntry(id: id, eventName: event, callback: listener))
    lock.unlock()
    return ListenerSubscription(remove: { [weak self] in
      self?.remove(id: id)
    })
  }

  func emit<T>(event: String, invoke: (T) throws -> Void) {
    lock.lock()
    let snapshot = listeners.filter { $0.eventName == event }
    lock.unlock()
    for entry in snapshot {
      if let callback = entry.callback as? T {
        try? invoke(callback)
      }
    }
  }

  func clearAll() {
    lock.lock()
    listeners.removeAll()
    lock.unlock()
  }

  func hasListeners(event: String) -> Bool {
    lock.lock()
    defer { lock.unlock() }
    return listeners.contains { $0.eventName == event }
  }

  private func remove(id: UUID) {
    lock.lock()
    listeners.removeAll { $0.id == id }
    lock.unlock()
  }
}
